runtime/common/src/fees.rs

```diff
@@ -0,0 +1,437 @@
+                    };
+                    let meta: AssetMetadata<Balance, CustomMetadata> = AssetMetadata {
+                        decimals: 10,
+                        name: "NoneLocationToken".into(),
+                        symbol: "NONE".into(),
+                        existential_deposit: ExistentialDeposit::get(),
+                        location: None,
+                        additional: custom_metadata,
+                    };
+                    let non_location_token = Asset::ForeignAsset(1);
+
+                    assert_ok!(AssetRegistry::register_asset(RuntimeOrigin::root(), meta, Some(non_location_token)));
+
+                    assert_noop!(get_fee_factor_and_base(non_location_token), TransactionValidityError::Invalid(InvalidTransaction::Custom(2u8)));
+                }
+            });
+        }
+
+        #[test]
+        fn get_fee_factor_and_base_decimals_overflow() {
+            let mut t: sp_io::TestExternalities =
+                frame_system::GenesisConfig::default().build_storage::<Runtime>().unwrap().into();
+            t.execute_with(|| {
+                #[cfg(feature = "parachain")]
+                {
+                    let custom_metadata = CustomMetadata {
+                        xcm: XcmMetadata { fee_factor: Some(102_993) },
+                        ..Default::default()
+                    };
+                    let meta: AssetMetadata<Balance, CustomMetadata> = AssetMetadata {
+                        decimals: u32::MAX,
+                        name: "OverflowCurrency".into(),
+                        symbol: "OVER".into(),
+                        existential_deposit: ExistentialDeposit::get(),
+                        location: Some(xcm::VersionedMultiLocation::V1(xcm::latest::MultiLocation::parent())),
+                        additional: custom_metadata,
+                    };
+                    let overflow_currency = Asset::ForeignAsset(1);
+
+                    assert_ok!(AssetRegistry::register_asset(RuntimeOrigin::root(), meta, Some(overflow_currency)));
+
+                    assert_noop!(get_fee_factor_and_base(overflow_currency), TransactionValidityError::Invalid(InvalidTransaction::Custom(4u8)));
+                }
+            });
+        }
+
+        #[test]
+        fn fee_multiplier_can_grow_from_zero() {
+            let minimum_multiplier = MinimumMultiplier::get();
+            let target = TargetBlockFullness::get()
+                * RuntimeBlockWeights::get().get(DispatchClass::Normal).max_total.unwrap();
+            // if the min is too small, then this will not change, and we are doomed forever.
+            // the weight is 1/100th bigger than target.
+            run_with_system_weight(target * 101 / 100, || {
+                let next = SlowAdjustingFeeUpdate::<Runtime>::convert(minimum_multiplier);
+                assert!(next > minimum_multiplier, "{:?} !>= {:?}", next, minimum_multiplier);
+            })
+        }
```

Comment:
This looks like something from Parity, so I think we should add the appropriate copyright.

Reply:
90027d7527d235139c47cf5fc24cf3928ba7762b